lib/restforce/db/associator.rb

```diff
@@ -0,0 +1,104 @@
+module Restforce
+
+  module DB
+
+    # Restforce::DB::Associator is responsible for determining when one or more
+    # associations have been updated to point to a new Salesforce/database
+    # record, and propagate the modification to the opposite system when this
+    # occurs.
+    class Associator
+
+      # Public: Initialize a new Restforce::DB::Associator.
+      #
+      # mapping - A Restforce::DB::Mapping instance.
+      # runner  - A Restforce::DB::Runner instance.
+      def initialize(mapping, runner = Runner.new)
+        @mapping = mapping
+        @runner = runner
+      end
+
+      # Public: Run the re-association process, pulling in records from
+      # Salesforce and the database to determine the most recently attached
+      # association, then propagating the change between systems.
+      #
+      # Returns nothing.
+      def run
+        return if associations.empty?
+
+        @runner.run(@mapping) do |run|
+          run.salesforce_instances { |instance| verify_associations(instance) }
+          run.database_instances { |instance| verify_associations(instance) }
+        end
+      end
+
+      private
+
+      # Internal: Ensure integrity between the lookup columns in Salesforce and
+      # the synchronized records in the database.
+      #
+      # instance - A Restforce::DB::Instances::Base.
+      #
+      # Returns nothing.
+      def verify_associations(instance)
+        database_instance, salesforce_instance =
+          case instance
```

Comment:
I'm not in love with this `case` statement. I don't think we should change it (because pushing the logic into the `Instance` subclasses doesn't actually seem better). I just wanted to gripe.